### PR TITLE
fix: auto register Vaadin servlet even if there is no routes

### LINKED ISSUE #4850: Validate automatic creation of main application servlet
After DynamicRoutes we should validate if a main application servlet should possibly be created even if we have no registered routes in the GlobalRouteRegistry.

The current implementation will not try to create a main application servlet if no Routes have been registered. With dynamic routes there is a possibility of adding the routes later or only to the session scope.

### PR DESCRIPTION
**Pull request by denis-anisimov**
fixes #4850

## Description

Please list all relevant dependencies in this section and provide summary of the change, motivation and context.

Fixes # (issue)

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [ ] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [ ] I have added a description following the guideline.
- [ ] The issue is created in the corresponding repository and I have referenced it.
- [ ] I have added tests to ensure my change is effective and works as intended.
- [ ] New and existing tests are passing locally with my change.
- [ ] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

**Review by caalador — DISMISSED** [review]

**Comment by caalador** [comment]
Conflicting files
flow-server/src/test/java/com/vaadin/flow/server/startup/ServletDeployerTest.java

**Comment by caalador** [comment]
```
/opt/agent/work/a69251da2b2bb9e9/flow-server/src/test/java/com/vaadin/flow/server/startup/ServletDeployerTest.java:[118,37] method getContextEvent in class com.vaadin.flow.server.startup.ServletDeployerTest cannot be applied to given types;
  required: javax.servlet.ServletRegistration[]
  found: boolean,boolean,javax.servlet.ServletRegistration
  reason: varargs mismatch; boolean cannot be converted to javax.servlet.ServletRegistration
```

**Comment by denis-anisimov** [comment]
done

File: flow-server/src/test/java/com/vaadin/flow/server/startup/ServletDeployerTest.java [hunk 1]
```diff
@@ -381,25 +320,13 @@ private ServletRegistration getServletRegistration(String servletName,
...
+    private void assertservletsAreNotRegistered_registeredServlets(
             boolean compatibilityMode) throws Exception {
-        deployer.contextInitialized(getContextEvent(false, false,
+        deployer.contextInitialized(getContextEvent(
                 getServletRegistration("testServlet", TestServlet.class,
                         singletonList("/test/*"), Collections.emptyMap())));
 
-        assertMappingsCount(0, 0);
+        assertMappingsCount(1, 1);
     }
```
Comment (caalador) [inline flow-server/src/test/java/com/vaadin/flow/server/startup/ServletDeployerTest.java:330]: Doesn't this now assert that a servlet is registered and not no servlets registered. 
Also the compatibilityMode boolean doesn't do anything in the method so it seems a bit non functional, and makes
`registeredServlets_compatibilityMode_servletsAreNotRegistered` and `registeredServlets_servletsAreNotRegistered` the same test, so I'm not certain what these actually test.

**Review by caalador — CHANGES_REQUESTED** [review]

File: flow-server/src/test/java/com/vaadin/flow/server/startup/ServletDeployerTest.java (hunk 1 above)
Comment (denis-anisimov) [reply flow-server/src/test/java/com/vaadin/flow/server/startup/ServletDeployerTest.java:330]: Yeah, a number of confusions.

compatibility mode is not checked anymore anyhow. This is a leftover from previous versions where `"frontend"` URL mapping servlet was registered as well.

The test has a precondition with non vaddin servlets and it checks that Vaadin servlet is registered. And previously the precondition was also no routes and no web components , that's why it checked that Vaadin servlet is not registered.
Now it should check that it's registered.
I've changed the names.

**Review by caalador — APPROVED** [review]

**Pull request merged**